### PR TITLE
Fix default handling for value types when converter based on interface

### LINKED ISSUE #42237: Custom struct not ignored during serialization when default
### Description

In a [test project of mine](https://github.com/Turnerj/SchemaNet_SystemTextJson/blob/master/SchemaNet_SystemTextJson/Program.cs) in my [ongoing effort to convert from Newtonsoft.Json to System.Text.Json](https://github.com/RehanSaeed/Schema.NET/pull/100), I found an issue relating to the handling of default values during serialization.

The test project has a class like the following:

```csharp
public class CustomType : Thing, IThing
{
	[JsonPropertyName("@type")]
	public override string Type => "CustomType";

	[JsonPropertyName("number")]
	[JsonConverter(typeof(ValuesJsonConverter))]
	public OneOrMany<int?> Number { get; set; }

	[JsonPropertyName("uri")]
	[JsonConverter(typeof(ValuesJsonConverter))]
	public Values<string, Uri> Uri { get; set; }
}
```

The types `OneOrMany<T>` and `Values<T1,T2>` are both `struct` and I believe this is where the problem lies (as changing them to `class` avoids this issue altogether).

**As Object**
```csharp
new CustomType
{
	Number = 123
}
```
**The JSON output by S.T.J**
```json
{"@type":"CustomType","number":123,"uri":null,"@context":"https://schema.org"}
```

**The desired JSON output**
```json
{"@type":"CustomType","number":123,"@context":"https://schema.org"}
```

**Serializer Options**
```csharp
new JsonSerializerOptions
{
	DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingDefault,
	Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
}
```

Now, I know where the `null` value comes from - [it comes from my `ValuesJsonConverter`](https://github.com/Turnerj/SchemaNet_SystemTextJson/blob/60efc1a8df655a77679a83a3e0abdedb91aaf986/SchemaNet_SystemTextJson/Schema.NET/ValuesJsonConverter.cs#L113) - however given that the property `Uri` is a default value (`default(Values<string,Uri>)`), my understanding is that it never should have got to my converter.

At the point my converter knows it is a default value, the property name has already been written out to the JSON so I can't do anything about it (writing nothing causes invalid JSON).

### Configuration

- .NET Core 3.1
- System.Text.Json 5.0.0-rc.1.20451.14
- Windows 10 Home (10.0.18362 Build 18362)

### Regression?

Unable to determine if this is a regression due to dependency on other fixes in RC1.

### Other information

You can see a working example of the issue here: https://github.com/Turnerj/SchemaNet_SystemTextJson/blob/master/SchemaNet_SystemTextJson/Program.cs

One known workaround for this is change the `OneOrMany<T>` and `Values<T1,T2>` to be a `class` instead of a `struct`. That isn't ruled out at this point though I think this is still a legitimate bug regardless which direction I go in my project.

### PR DESCRIPTION
**Pull request by steveharter**
Fixes https://github.com/dotnet/runtime/issues/42237

When a custom converter is based on an interface and a corresponding property is based on a value type that implements the interface, default handling was not working. This PR fixes that.

Some local benchmarks showed no regression for the happy path that doesn't hit this edge case.

### CODE REVIEW

File: src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs [hunk 1]
```diff
@@ -2367,5 +2367,152 @@ private struct StructWithBadIgnoreAttribute
...
+            Assert.Equal("0", json);
+        }
+
+        private struct MyValueTypeWithBoxedPrimitive
+        {
+            public object BoxedPrimitive { get; set; }
+        }
+
+        [Fact]
+        public static void JsonIgnoreCondition_WhenWritingDefault_OnBoxedPrimitive()
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs:2494]: This was passing before, just added the test.

File: src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs [hunk 2]
```diff
@@ -2367,5 +2367,152 @@ private struct StructWithBadIgnoreAttribute
...
+            };
+
+            // Verify not ignored; MyValueType is not null.
+            Assert.True(EqualityComparer<MyValueTypeWithProperties>.Default.Equals(default, obj.Value));
+            json = JsonSerializer.Serialize(obj, options);
+            Assert.Equal("{\"Value\":100}", json);
+        }
+
+        [Fact]
+        public static void JsonIgnoreCondition_WhenWritingDefault_OnRootTypes()
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs:2461]: This was passing before, just added the test.

File: src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs [hunk 3]
```diff
@@ -2367,5 +2367,152 @@ private struct StructWithBadIgnoreAttribute
...
+
+        private class MyCustomConverter : JsonConverter<IUseCustomConverter>
+        {
+            public override bool CanConvert(Type typeToConvert)
+            {
+                return typeof(IUseCustomConverter).IsAssignableFrom(typeToConvert);
+            }
+
+            public override IUseCustomConverter Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) =>
+                throw new NotImplementedException();
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs:2388]: Since `IgnoreDefaultValues` is applied to deserialization, can we have one or two assertions on whether this is called?

File: src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs [hunk 4]
```diff
@@ -2367,5 +2367,152 @@ private struct StructWithBadIgnoreAttribute
...
+            Assert.True(EqualityComparer<MyValueTypeWithProperties>.Default.Equals(default, obj.Value));
+            json = JsonSerializer.Serialize(obj);
+            Assert.Equal("{\"Value\":100}", json);
+
+            var options = new JsonSerializerOptions
+            {
+                DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
+            };
+
+            // Verify not ignored; MyValueType is not null.
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs:2454]: > MyValueType is not null.

What entity (property/class...) is this referring to?

File: src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs [hunk 5]
```diff
@@ -2367,5 +2367,152 @@ private struct StructWithBadIgnoreAttribute
...
+            // Baseline without custom options.
+            json = JsonSerializer.Serialize(obj);
+            Assert.Equal("{\"BoxedPrimitive\":0}", json);
+
+            var options = new JsonSerializerOptions
+            {
+                DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingDefault
+            };
+
+            // No check on the boxed object to see the value type is a default value (0 in this case).
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs:2509]: Nice test. I assume it's the same behavior when the property is an interface and not object:

```cs
public class MyClass
{
    [JsonIgnore(JsonIgnoreCondition.WhenWritingDefault)]
    public IInterface MyProp { get; set; }
}

public interface IInterface { }

public struct MyStruct : Interface { }

var obj = new MyClass {  MyProp = new MyStruct() };
string serialized = JsonSerializer.Serialize(obj); // {"MyProp":{}}
```

I assume we wouldn't ever do `default` check based on the runtime value even with extended polymorphic support since the check for `default` would be expensive (e.g new-ing up a `GenericMethodHolder` for every property).

File: src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs (hunk 4 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs:2454]: Shoud be `MyValueTypeWithProperties`

File: src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs (hunk 3 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs:2388]: How is `IgnoreDefaultValues` used during deserialization? My understanding is only the older `IgnoreNullValues` affects deserialization.

File: src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs (hunk 5 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs:2509]: Correct. I'll add the test

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/GenericMethodHolder.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    /// <summary>
+    /// Generic methods for {T}.
+    /// </summary>
+    internal sealed class GenericMethodHolder<T> : GenericMethodHolder
+    {
+        public override bool IsDefaultValue(object value)
+        {
+            // For performance, we only want to call this method for value types.
+            // Reference types should be checked againt 'null' before calling this method.
+            Debug.Assert(value.GetType().IsValueType);
```
[the remark this answers is not in the document]
Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/GenericMethodHolder.cs]: Sure. Thanks

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs [hunk 7]
```diff
@@ -100,7 +105,10 @@ internal sealed class JsonPropertyInfo<T> : JsonPropertyInfo
             }
 
             _converterIsExternalAndPolymorphic = !converter.IsInternalConverter && DeclaredPropertyType != converter.TypeToConvert;
-            GetPolicies(ignoreCondition, parentTypeNumberHandling, defaultValueIsNull: Converter.CanBeNull);
+            _propertyTypeCanBeNull = !declaredPropertyType.IsValueType || Nullable.GetUnderlyingType(declaredPropertyType) != null;
```
[the remark this answers is not in the document]
Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs]: Thanks; changed it. I believe this occurs in 1-2 other locations (to follow up later in separate PR).

File: src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs (hunk 5 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs]: Also note that `GenericMethodHolder ` would only in theory be newed up for every property type, not every property, since it is cached on the ClassInfo for that property type.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs [hunk 8]
```diff
@@ -131,18 +139,39 @@ public override bool GetMemberAndWriteJson(object obj, ref WriteStack state, Utf
...
+                        {
+                            return true;
+                        }
+                    }
+                    else
+                    {
+                        // Use a late-bound call to EqualityComparer<DeclaredPropertyType>.
+                        if (RuntimeClassInfo.GenericMethods.IsDefaultValue(value))
+                        {
+                            Debug.Assert(RuntimeClassInfo.Type == DeclaredPropertyType);
```
[the remark this answers is not in the document]
Comment (steveharter) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs]: Sure. Thanks

File: src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs (hunk 3 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs:2388]: In either case, I added a test for `IgnoreNullValues`.

Comment (layomia) [reply src/libraries/System.Text.Json/tests/Serialization/PropertyVisibilityTests.cs:2388]: Yes I meant `IgnoreNullValues`, thanks.

**Review by layomia — APPROVED** [review]
LGTM. Would like to present to ship-room for .NET 5 consideration (incomplete scenario for ignoring default values feature new in 5.0).

**Comment by steveharter** [comment]
Runtime failure in mono unrelated:
```
##[error]Extraction failed for file: F:\workspace\_work\1\s\__download__\libraries_test_assets_Windows_NT_x64_Debug\libraries_test_assets_Windows_NT_x64_Debug.zip 
code: 2 
```
`
Publishing build artifacts failed with an error: Not found PathtoPublish: F:\workspace\_work\1\s\artifacts\log\Debug
`
